src/components/tl/ucp/alltoall/alltoall_onesided.c

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "alltoall.h"
+#include "core/ucc_progress_queue.h"
+#include "utils/ucc_math.h"
+#include "tl_ucp_sendrecv.h"
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_start(ucc_coll_task_t *ctask)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team = TASK_TEAM(task);
+    ucc_status_t       status;
+
+    task->barrier.phase = 0;
+
+    task->super.super.status = UCC_INPROGRESS;
+    status = ucc_tl_ucp_alltoall_onesided_progress(&task->super);
+    if (UCC_INPROGRESS == status) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+    task->super.super.status = status;
+    ucc_task_complete(ctask);
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_alltoall_onesided_progress(ucc_coll_task_t *ctask)
+{
+    ucc_tl_ucp_task_t *task   = ucc_derived_of(ctask, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team   = TASK_TEAM(task);
+    ptrdiff_t          src    = (ptrdiff_t)TASK_ARGS(task).src.info.buffer;
+    ptrdiff_t          dest   = (ptrdiff_t)TASK_ARGS(task).dst.info.buffer;
+    size_t             nelems = TASK_ARGS(task).src.info.count;
+    ucc_rank_t         mype   = team->rank;
+    ucc_rank_t         npes   = team->size;
+    int *              phase  = &task->barrier.phase;
+    ucc_status_t       status = UCC_INPROGRESS;
+    long *             pSync  = TASK_ARGS(task).pSync;
+    ucc_rank_t         peer;
+
+    if (*phase == 1) {
+        goto wait;
+    }
+
+    dest             = dest + mype * nelems;
+    ucc_rank_t start = (mype + 1) % npes; 
+    ucc_tl_ucp_put_nb((void *)(src + start * nelems), (void *)dest, nelems,
+                      start, team, task);
+    ucc_tl_ucp_atomic_inc(pSync, start, team, task);
+    ucc_tl_ucp_ep_flush(start, team, task);
+    
+    for (peer = (start + 1) % npes; peer != start; peer = (peer + 1) % npes) {
+        ucc_tl_ucp_put_nb((void *)(src + peer * nelems), (void *)dest, nelems,
+                          peer, team, task);
+        ucc_tl_ucp_atomic_inc(pSync, peer, team, task);
+        ucc_tl_ucp_ep_flush(peer, team, task);
```

Comment:
Do you really need a flush after every put?

Reply:
@manjugv the flush can be there or after the loop, but there should be a flush to guarantee completion of operation...